File: lvc-mac/lvc-mac/ConnectionManager.swift
//
//  ConnectionManager.swift
//  lvc-mac
//
//  Created by Zack Clark-Kington on 17/12/2017.
//  Copyright © 2017 Zack Clark-Kington. All rights reserved.
//

import Foundation
import Cocoa

class ConnectionManager{
    private var host : String? = nil
    private var port : String? = nil
    

    init(host: String, port: String) {
        self.host = host
        self.port = port
        
        self.testConnection()
    }
    
    private func makeRequest(requestEndpoint: String) -> Any?{
        print(requestEndpoint)
        let url = URL(string: "http://" + self.host! + ":" + self.port! + requestEndpoint)
        let request = URLRequest(url: url!)
        let response:AutoreleasingUnsafeMutablePointer<URLResponse?>?=nil
        
        do{
            let responseData = try NSURLConnection.sendSynchronousRequest(request, returning: response)
            return responseData
        } catch(let e){
            let alert = NSAlert.init(error: e)
            alert.runModal()
        }
        
        return nil
    }
    
    public func testConnection() -> Bool{
        let response = makeRequest(requestEndpoint: "/api/registerClient")
        return !(response == nil)
    }
    
    public func getVersionedDocuments() -> NSMutableDictionary{
        let response : Data = makeRequest(requestEndpoint: "/api/getFileList") as! Data
        do{
            return try JSONSerialization.jsonObject(with: response, options: JSONSerialization.ReadingOptions.mutableContainers) as! NSMutableDictionary
        } catch {
            return NSMutableDictionary.init(dictionary: [String : String](), copyItems: false)
        }
    }
    
    public func getNumberOfRevisions(fp: String) -> Int{
        let requestEndpoint = "/api/getNumberRevisions/" + fp
        let responseData : Data = makeRequest(requestEndpoint: requestEndpoint) as! Data
        do{
            let jsonDict = try JSONSerialization.jsonObject(with: responseData, options: JSONSerialization.ReadingOptions.mutableContainers) as! NSMutableDictionary
            return Int(jsonDict.value(forKey: "numRevisions") as! NSNumber)
        } catch {
            return 0
        }
    }
    
    public func getRevisionByIndex(fp: String, index: Int) -> String{
        let requestEndpoint = "/api/getRevisionByIndex/" + String(index) + "/" + fp
        let responseData : Data = makeRequest(requestEndpoint: requestEndpoint) as! Data
        do{
            let jsonDict = try JSONSerialization.jsonObject(with: responseData, options: JSONSerialization.ReadingOptions.mutableContainers) as! NSMutableDictionary
            return jsonDict.value(forKey: "content") as! String
        } catch {
            return ""
        }
    }
    
    public func getLatestRevision(fp: String) -> String{
        let latestRevisionIndex : Int = self.getNumberOfRevisions(fp: fp) - 1
        return getRevisionByIndex(fp: fp, index: latestRevisionIndex)
    }
}

File: electronapp/app/ConnectionManager.js
function ConnectionManager(host, port){
	this.host = host;
	this.port = port;

	this.makeRequest = function(requestEndpoint, cb){
		try{
			jQuery.get('http://' + this.host + ':' + this.port + '/' + requestEndpoint, function(r){
				cb($.parseJSON(r));
			});
		} catch(e) {
			cb(null);
		}
	};

	this.testConnection = function(cb){
		var response = this.makeRequest('api/registerClient', cb);
	}

	this.getVersionedFiles = function(cb){
		return this.makeRequest('api/getFileList', cb);
	}

	this.getNumberRevisions = function(fp){
		var response = this.makeRequest('api/getNumberRevisions/' + fp);
		return response.numRevisions;
	};

	this.getRevisionByIndex = function(index, fp, cb){
		var response = this.makeRequest('api/getRevisionByIndex/' + index + '/' + fp);
		return response.content;
	}

	this.getLatestRevision = function(fp, cb){
		var latestRevisionIndex = this.getNumberRevisions(fp) - 1;
		var response = this.getRevisionByIndex(latestRevisionIndex, fp);
		return response.content;
	}

	return this;
}